File: civ65/mapgen.py
"""Random map generation: continents, climate bands, features, resources, spawns."""
from . import hexgrid
from .world import WorldMap


def generate(rules, width, height, num_players, rng):
    wm = WorldMap(width, height)
    tiles = wm.tiles

    # --- landmass via random frontier growth
    land_target = int(0.42 * width * height)
    land = set()
    for _ in range(max(2, num_players // 2 + 1)):
        col = rng.randint(width // 5, width - 1 - width // 5)
        row = rng.randint(height // 5, height - 1 - height // 5)
        land.add(hexgrid.offset_to_axial(col, row))
    frontier = sorted(land)
    for _ in range(40 * width * height):
        if len(land) >= land_target:
            break
        if not frontier:
            frontier = [sorted(land)[rng.randrange(len(land))]]
        cur = frontier[rng.randrange(len(frontier))]
        nbs = [n for n in hexgrid.neighbors(cur) if n in tiles and n not in land]
        if not nbs:
            frontier.remove(cur)
            continue
        nxt = nbs[rng.randrange(len(nbs))]
        land.add(nxt)
        frontier.append(nxt)
        if rng.random() < 0.25:
            frontier.remove(cur)

    land_list = sorted(land)

    # --- climate bands
    def latitude(pos):
        _, row = hexgrid.axial_to_offset(*pos)
        return abs(2.0 * row / max(1, height - 1) - 1.0)

    for pos in land_list:
        v = latitude(pos) + rng.uniform(-0.15, 0.15)
        t = tiles[pos]
        if v > 0.85:
            t.terrain = "snow"
        elif v > 0.62:
            t.terrain = "tundra"
        elif v < 0.30:
            t.terrain = rng.choices(["desert", "plains", "grassland"], [0.25, 0.40, 0.35])[0]
        else:
            t.terrain = rng.choices(["grassland", "plains", "desert"], [0.50, 0.42, 0.08])[0]

    # --- mountain chains, hills
    for _ in range(max(2, len(land_list) // 40)):
        cur = land_list[rng.randrange(len(land_list))]
        for _ in range(rng.randint(2, 5)):
            tiles[cur].terrain = "mountain"
            nbs = [n for n in hexgrid.neighbors(cur) if n in land]
            if not nbs:
                break
            cur = nbs[rng.randrange(len(nbs))]
    for pos in land_list:
        if tiles[pos].terrain != "mountain" and rng.random() < 0.16:
            tiles[pos].hills = True

    # --- coast
    for pos, t in tiles.items():
        if t.terrain == "ocean" and any(n in land for n in hexgrid.neighbors(pos)):
            t.terrain = "coast"

    # --- features
    for pos in land_list:
        t = tiles[pos]
        if t.terrain == "mountain":
            continue
        r = rng.random()
        if t.terrain in ("grassland", "plains"):
            if latitude(pos) < 0.25 and r < 0.28:
                t.feature = "jungle"
            elif r < 0.20:
                t.feature = "forest"
            elif t.terrain == "grassland" and r > 0.97:
                t.feature = "marsh"
        elif t.terrain == "tundra" and r < 0.22:
            t.feature = "forest"
        elif t.terrain == "desert" and r < 0.05:
            t.feature = "oasis"

    # --- resources
    for pos in sorted(tiles):
        t = tiles[pos]
        if t.terrain == "mountain" or t.feature in ("oasis", "marsh"):
            continue
        if rng.random() < 0.13:
            valid = []
            for name, s in rules.resources.items():
                if s.get("feature"):
                    if t.feature in s["feature"]:
                        valid.append(name)
                elif t.terrain in s.get("terrain", []):
                    valid.append(name)
            if valid:
                t.resource = valid[rng.randrange(len(valid))]

    # --- spawns on the largest connected passable landmass
    passable = {p for p in land if tiles[p].terrain != "mountain"}
    largest = _largest_component(passable)
    cands = sorted(p for p in largest
                   if tiles[p].terrain in ("grassland", "plains") and not tiles[p].feature)
    if len(cands) < num_players:
        cands = sorted(largest)
    spawns = [cands[rng.randrange(len(cands))]]
    while len(spawns) < num_players:
        pool = [c for c in cands if c not in spawns]
        best = max(pool, key=lambda c: (min(hexgrid.distance(c, s) for s in spawns), c))
        spawns.append(best)
    for s in spawns:
        tiles[s].feature = None
        tiles[s].resource = None
    return wm, spawns


def _largest_component(cells):
    seen = set()
    best = set()
    for start in sorted(cells):
        if start in seen:
            continue
        comp = {start}
        stack = [start]
        while stack:
            cur = stack.pop()
            for n in hexgrid.neighbors(cur):
                if n in cells and n not in comp:
                    comp.add(n)
                    stack.append(n)
        seen |= comp
        if len(comp) > len(best):
            best = comp
    return best

File: rust/src/mapgen.rs
//! Map generation (mirrors civvis/mapgen.py).
use std::collections::BTreeSet;

use crate::rng::Rng;
use crate::rules::Rules;
use crate::world::WorldMap;
use crate::{hex, Pos};

pub fn generate(rules: &Rules, width: i32, height: i32, num_spawns: usize,
                rng: &mut Rng) -> (WorldMap, Vec<Pos>) {
    let mut wm = WorldMap::new(width, height);

    // --- landmass via random frontier growth
    let land_target = (0.42 * (width * height) as f64) as usize;
    let mut land: BTreeSet<Pos> = BTreeSet::new();
    for _ in 0..2.max(num_spawns / 2 + 1) {
        let col = rng.randint(width / 5, width - 1 - width / 5);
        let row = rng.randint(height / 5, height - 1 - height / 5);
        land.insert(hex::offset_to_axial(col, row));
    }
    let mut frontier: Vec<Pos> = land.iter().cloned().collect();
    for _ in 0..(40 * width * height) {
        if land.len() >= land_target {
            break;
        }
        if frontier.is_empty() {
            let idx = rng.below(land.len());
            frontier.push(*land.iter().nth(idx).unwrap());
        }
        let ci = rng.below(frontier.len());
        let cur = frontier[ci];
        let nbs: Vec<Pos> = hex::neighbors(cur)
            .into_iter()
            .filter(|n| wm.tiles.contains_key(n) && !land.contains(n))
            .collect();
        if nbs.is_empty() {
            frontier.remove(ci);
            continue;
        }
        let nxt = nbs[rng.below(nbs.len())];
        land.insert(nxt);
        frontier.push(nxt);
        if rng.chance(0.25) {
            if let Some(i) = frontier.iter().position(|p| *p == cur) {
                frontier.remove(i);
            }
        }
    }

    let land_list: Vec<Pos> = land.iter().cloned().collect();
    let latitude = |pos: Pos| -> f64 {
        let (_, row) = hex::axial_to_offset(pos.0, pos.1);
        (2.0 * row as f64 / (height - 1).max(1) as f64 - 1.0).abs()
    };

    // --- climate bands
    for pos in &land_list {
        let v = latitude(*pos) + rng.uniform(-0.15, 0.15);
        let t = wm.tiles.get_mut(pos).unwrap();
        t.terrain = if v > 0.85 {
            "snow".into()
        } else if v > 0.62 {
            "tundra".into()
        } else if v < 0.30 {
            ["desert", "plains", "grassland"][rng.weighted(&[0.25, 0.40, 0.35])].into()
        } else {
            ["grassland", "plains", "desert"][rng.weighted(&[0.50, 0.42, 0.08])].into()
        };
    }

    // --- mountain chains, hills
    for _ in 0..2.max(land_list.len() / 40) {
        let mut cur = land_list[rng.below(land_list.len())];
        let steps = rng.randint(2, 5);
        for _ in 0..steps {
            wm.tiles.get_mut(&cur).unwrap().terrain = "mountain".into();
            let nbs: Vec<Pos> = hex::neighbors(cur)
                .into_iter()
                .filter(|n| land.contains(n))
                .collect();
            if nbs.is_empty() {
                break;
            }
            cur = nbs[rng.below(nbs.len())];
        }
    }
    for pos in &land_list {
        let roll = rng.chance(0.16);
        let t = wm.tiles.get_mut(pos).unwrap();
        if t.terrain != "mountain" && roll {
            t.hills = true;
        }
    }

    // --- coast
    let coastal: Vec<Pos> = wm
        .tiles
        .iter()
        .filter(|(pos, t)| {
            t.terrain == "ocean" && hex::neighbors(**pos).iter().any(|n| land.contains(n))
        })
        .map(|(pos, _)| *pos)
        .collect();
    for pos in coastal {
        wm.tiles.get_mut(&pos).unwrap().terrain = "coast".into();
    }

    // --- features
    for pos in &land_list {
        let lat = latitude(*pos);
        let r = rng.f64();
        let t = wm.tiles.get_mut(pos).unwrap();
        if t.terrain == "mountain" {
            continue;
        }
        if t.terrain == "grassland" || t.terrain == "plains" {
            if lat < 0.25 && r < 0.28 {
                t.feature = Some("jungle".into());
            } else if r < 0.20 {
                t.feature = Some("forest".into());
            } else if t.terrain == "grassland" && r > 0.97 {
                t.feature = Some("marsh".into());
            }
        } else if t.terrain == "tundra" && r < 0.22 {
            t.feature = Some("forest".into());
        } else if t.terrain == "desert" && r < 0.05 {
            t.feature = Some("oasis".into());
        }
    }

    // --- resources
    let all_pos: Vec<Pos> = wm.tiles.keys().cloned().collect();
    for pos in all_pos {
        let (terrain, feature) = {
            let t = &wm.tiles[&pos];
            (t.terrain.clone(), t.feature.clone())
        };
        if terrain == "mountain"
            || feature.as_deref() == Some("oasis")
            || feature.as_deref() == Some("marsh")
        {
            continue;
        }
        if rng.chance(0.13) {
            let valid: Vec<String> = rules
                .resources
                .iter()
                .filter(|(_, s)| {
                    if !s.feature.is_empty() {
                        feature.as_ref().map(|f| s.feature.contains(f)).unwrap_or(false)
                    } else {
                        s.terrain.contains(&terrain)
                    }
                })
                .map(|(name, _)| name.clone())
                .collect();
            if !valid.is_empty() {
                let pick = valid[rng.below(valid.len())].clone();
                wm.tiles.get_mut(&pos).unwrap().resource = Some(pick);
            }
        }
    }

    // --- spawns on the largest connected passable landmass
    let passable: BTreeSet<Pos> = land
        .iter()
        .filter(|p| wm.tiles[p].terrain != "mountain")
        .cloned()
        .collect();
    let largest = largest_component(&passable);
    let mut cands: Vec<Pos> = largest
        .iter()
        .filter(|p| {
            let t = &wm.tiles[p];
            (t.terrain == "grassland" || t.terrain == "plains") && t.feature.is_none()
        })
        .cloned()
        .collect();
    if cands.len() < num_spawns {
        cands = largest.iter().cloned().collect();
    }
    cands.sort();
    let mut spawns = vec![cands[rng.below(cands.len())]];
    while spawns.len() < num_spawns {
        let pool: Vec<Pos> = cands.iter().filter(|c| !spawns.contains(c)).cloned().collect();
        if pool.is_empty() {
            break;
        }
        let best = *pool
            .iter()
            .max_by_key(|c| {
                let d = spawns.iter().map(|s| hex::distance(**c, *s)).min().unwrap();
                (d, **c)
            })
            .unwrap();
        spawns.push(best);
    }
    for s in &spawns {
        let t = wm.tiles.get_mut(s).unwrap();
        t.feature = None;
        t.resource = None;
    }
    (wm, spawns)
}

fn largest_component(cells: &BTreeSet<Pos>) -> BTreeSet<Pos> {
    let mut seen: BTreeSet<Pos> = BTreeSet::new();
    let mut best: BTreeSet<Pos> = BTreeSet::new();
    for start in cells {
        if seen.contains(start) {
            continue;
        }
        let mut comp: BTreeSet<Pos> = BTreeSet::new();
        comp.insert(*start);
        let mut stack = vec![*start];
        while let Some(cur) = stack.pop() {
            for n in hex::neighbors(cur) {
                if cells.contains(&n) && !comp.contains(&n) {
                    comp.insert(n);
                    stack.push(n);
                }
            }
        }
        seen.extend(comp.iter().cloned());
        if comp.len() > best.len() {
            best = comp;
        }
    }
    best
}
